swiftpkg/internal/swift_registry_package.bzl

```diff
@@ -0,0 +1,74 @@
+"""Implementation for `swift_registry_package`."""
+
+load("@bazel_skylib//lib:dicts.bzl", "dicts")
+load("@bazel_tools//tools/build_defs/repo:utils.bzl", "update_attrs")
+load(":pkg_ctxs.bzl", "pkg_ctxs")
+load(":repo_rules.bzl", "repo_rules")
+load(":repository_files.bzl", "repository_files")
+
+def _remove_bazel_files(repository_ctx, directory):
+    files = ["BUILD.bazel", "BUILD", "WORKSPACE", "WORKSPACE.bazel"]
+    for file in files:
+        repository_files.find_and_delete_files(repository_ctx, directory, file)
+
+def _swift_registry_package_impl(repository_ctx):
+    directory = str(repository_ctx.path("."))
+    env = repo_rules.get_exec_env(repository_ctx)
+    repo_rules.check_spm_version(repository_ctx, env = env)
+
+    url = repository_ctx.attr.url
+
+    # Works for my case, but the prefix can actually be anything, so this is not sufficient for general use.
+    # See `archive-source` command docs:
+    # https://github.com/apple/swift-evolution/blob/main/proposals/0292-package-registry-service.md#archive-source-subcommand
+    components = url.split("/")
```

Comment:
What do we need to make this work more generally?

Reply:
> What do we need to make this work more generally?

Something like `stripPrefix` that doesn't require you to know the prefix in advance. It could be something along the lines of:
1. download and extract into a temp directory
2. sanity check there is only one directory in the extracted file location
3. move all the files from that one directory to the `repository_ctx` root (where `download_and_extract` would have put them by default)

Not too difficult to script, but I don't know if there's a way to do it without shelling out to commands.

Reply:
I scripted out a more general solution.